src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/PkcsHelpers.cs

```diff
@@ -110,44 +111,56 @@ public static void RemoveAt<T>(ref T[] arr, int idx)
             arr = tmp;
         }
 
-        public static T[] NormalizeSet<T>(
-            T[] setItems,
+        public static AttributeAsn[] NormalizeAttributeSet(
+            AttributeAsn[] setItems,
             Action<byte[]> encodedValueProcessor=null)
         {
-            AsnSet<T> set = new AsnSet<T>
-            {
-                SetData = setItems,
-            };
-
-            AsnWriter writer = AsnSerializer.Serialize(set, AsnEncodingRules.DER);
-            byte[] normalizedValue = writer.Encode();
-            set = AsnSerializer.Deserialize<AsnSet<T>>(normalizedValue, AsnEncodingRules.DER);
+            byte[] normalizedValue;
 
-            if (encodedValueProcessor != null)
+            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.DER))
             {
-                encodedValueProcessor(normalizedValue);
+                writer.PushSequence();
+                writer.PushSetOf();
+                foreach (AttributeAsn item in setItems)
+                {
+                    item.Encode(writer);
+                }
+                writer.PopSetOf();
+                writer.PopSequence();
+                normalizedValue = writer.Encode();
+                if (encodedValueProcessor != null)
+                {
+                    encodedValueProcessor(normalizedValue);
+                }
             }
 
-            return set.SetData;
+            AsnReader reader = new AsnReader(normalizedValue, AsnEncodingRules.DER);
+            AsnReader sequenceReader = reader.ReadSequence();
+            AsnReader setReader = sequenceReader.ReadSetOf();
+            List<AttributeAsn> decodedSet = new List<AttributeAsn>(setItems.Length);
+            while (setReader.HasData)
+            {
+                AttributeAsn.Decode(setReader, out AttributeAsn item);
+                decodedSet.Add(item);
+            }
+            return decodedSet.ToArray();
```

Comment:
Since the number of items out is the number of items in, `decodedSet` could just be an array to avoid the re-copy at the end.